File: DFS和BFS/bfs 超级源点 最短的桥.go
package DFS和BFS

import "math"

type xx struct {
	row          int
	col          int
	bridgeLength int //桥的长度
}

func shortestBridge(grid [][]int) int {
	m := len(grid)
	n := len(grid[0])
	var queue []*xx
	var dfs func(i, j int)
	dfs = func(i, j int) {
		if i < 0 || j < 0 || i >= m || j >= n || grid[i][j] != 1 {
			return
		}
		grid[i][j] = 114514
		queue = append(queue, &xx{i, j, 0})
		dfs(i-1, j)
		dfs(i+1, j)
		dfs(i, j-1)
		dfs(i, j+1)
	}
	//通过dfs和这个for循环构建超级源点
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == 1 {
				dfs(i, j)
				goto label
			}
		}
	}
label:
	for len(queue) != 0 {
		QueueLengthOfThisLayer := len(queue)
		for i := 0; i < QueueLengthOfThisLayer; i++ {
			cur := queue[0]
			queue = queue[1:]
			row := cur.row
			col := cur.col
			if row-1 >= 0 {
				if grid[row-1][col] == 1 {
					return cur.bridgeLength
				} else if grid[row-1][col] == 0 {
					grid[row-1][col] = 114514
					queue = append(queue, &xx{row - 1, col, cur.bridgeLength + 1})
				}
			}
			if row+1 < m {
				if grid[row+1][col] == 1 {
					return cur.bridgeLength
				} else if grid[row+1][col] == 0 {
					grid[row+1][col] = 114514
					queue = append(queue, &xx{row + 1, col, cur.bridgeLength + 1})
				}
			}
			if col-1 >= 0 {
				if grid[row][col-1] == 1 {
					return cur.bridgeLength
				} else if grid[row][col-1] == 0 {
					grid[row][col-1] = 114514
					queue = append(queue, &xx{row, col - 1, cur.bridgeLength + 1})
				}
			}
			if col+1 < n {
				if grid[row][col+1] == 1 {
					return cur.bridgeLength
				} else if grid[row][col+1] == 0 {
					grid[row][col+1] = 114514
					queue = append(queue, &xx{row, col + 1, cur.bridgeLength + 1})
				}
			}
		}
	}
	//不可到达的代码
	return math.MaxInt32
}

File: DFS和BFS/BFS_超级源点_最短的桥.java
package DFS和BFS;

import java.util.ArrayDeque;
import java.util.Deque;

class Point {
    int x;
    int y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

public class BFS_超级源点_最短的桥 {
    int[][] grid;
    int n;
    Deque<Point> queue = new ArrayDeque<Point>();

    public int shortestBridge(int[][] grid) {
        this.grid = grid;
        n = grid[0].length;
        var foundFirstGrid = false;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    foundFirstGrid = true;
                    dfs(i, j);
                    break;
                }
            }
            if (foundFirstGrid) break;
        }
        int distance = 0;
        int[][] dirs = new int[][]{{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
        while (!queue.isEmpty()) {
            int x = queue.size();
            for (int i = 0; i < x; i++) {
                var cur = queue.poll();
                assert cur != null;
                for (int[] dir : dirs) {
                    var nextx = cur.x + dir[0];
                    var nexty = cur.y + dir[1];
                    if (nextx >= 0 && nexty >= 0 && nextx < n && nexty < n) {
                        if (grid[nextx][nexty] == 1) {
                            return distance;
                        } else if (grid[nextx][nexty] == 0) {
                            queue.add(new Point(nextx, nexty));
                            grid[nextx][nexty] = 2;
                        }
                    }
                }
            }
            distance++;
        }
        return -1;
    }

    public void dfs(int i, int j) {
        if (i < 0 || j < 0 || i >= n || j >= n || grid[i][j] != 1) return;
        grid[i][j] = 2;
        queue.add(new Point(i, j));
        dfs(i - 1, j);
        dfs(i, j - 1);
        dfs(i + 1, j);
        dfs(i, j + 1);
    }
}
